sweep/sweeper.go

```diff
@@ -30,13 +67,623 @@ type UtxoSweeperConfig struct {
+		// sweeps from happening in fast succession. The first sweep
+		// triggered by the maximum allowed number of inputs and the
+		// second because of the timer.
+		log.Debugf("No inputs remaining, disable timer")
+		s.timer = nil
+		return
+	}
+
+	if s.timer == nil {
+		// Start sweep timer to create opportunity for more inputs to be
+		// added before a tx is constructed.
+		log.Debugf("Start timer")
+		s.timer = s.cfg.NewBatchTimer()
+	} else {
+		log.Debugf("Timer still ticking")
+	}
+}
+
+func (s *UtxoSweeper) signalAndRemove(outpoint *wire.OutPoint, result Result) {
+	pendInput := s.pendingInputs[*outpoint]
+	listeners := pendInput.listeners
+
+	if result.Err == nil {
+		log.Debugf("Dispatching sweep success for %v to %v listeners",
+			outpoint, len(listeners))
+	} else {
+		log.Debugf("Dispatching sweep error for %v to %v listeners: %v",
+			outpoint, len(listeners), result.Err)
+	}
+
+	for _, resultChan := range listeners {
+		resultChan <- result
+	}
+
+	// Cancel spend notification with chain notifier. This is not necessary
+	// in case of a success, except for that a reorg could still happen.
+	if pendInput.ntfnRegCancel != nil {
+		pendInput.ntfnRegCancel()
+	}
+
+	// Inputs are no longer pending after result has been sent.
+	delete(s.pendingInputs, *outpoint)
+}
+
+// sweepAll goes through all pending inputs and constructs sweep transactions,
+// each up to the configured maximum number of inputs. Negative yield inputs are
+// skipped. Transactions with an output below the dust limit are not published.
+// Those inputs remain pending and will be bundled with future inputs if
+// possible.
+//
+// Returns a boolean indicating whether there are sweepable inputs remaining.
+// This can happen when fullOnly is true and there are less sweepable inputs
+// then the configured maximum number per tx.
+func (s *UtxoSweeper) sweepAll(fullOnly bool) (bool, error) {
+	// Retrieve fee estimate for input filtering and final tx fee
+	// calculation.
+	satPerKW, err := s.cfg.Estimator.EstimateFeePerKW(
+		s.cfg.SweepTxConfTarget)
+	if err != nil {
+		return false, err
+	}
+
+	// Retrieve current height to properly set tx locktime.
+	_, currentHeight, err := s.cfg.ChainIO.GetBestBlock()
+	if err != nil {
+		return false, err
+	}
+
+	// Filter for inputs that need to be swept.
+	var sweepableInputs []*pendingInput
+	for _, input := range s.pendingInputs {
+		// Skip inputs that are already part of a sweep tx.
+		if input.state != stateNew {
+			continue
+		}
+
+		sweepableInputs = append(sweepableInputs, input)
+	}
+
+	// Sort input by value. We will start constructing sweep transactions
+	// starting with the highest value inputs. This is to prevent the
+	// construction of a sweep tx with an output below the dust limit,
+	// causing the sweep process to stop, while there are still higher value
+	// inputs sweepable.
+	sort.Slice(sweepableInputs, func(i, j int) bool {
+		return sweepableInputs[i].input.SignDesc().Output.Value >
```

Comment:
Been thinking about a slightly different sorting function, let me know your thoughts on this proposal.

At the time we try to sort the inputs, we know:
 - `r`, sweep fee rate in sat/kw (same one used in sweep tx weight estimator)
 - `v` input value in sat
 - `w` worst-case witness weight (including witness script)

The goal is to order the inputs, such that for any pair of indeces (i, j) for i < j, adding input j results less expected yield than input i. Some observations:
 - all input types contribute equally to non-witness data
 - input types contribute distinct-but-known worst-case amounts to witness data
 - the contribution towards the numinputs varint is a negative delta function at a specific index, and if we assume that the inputs are in the sorted order described above, then no other ordering would be able to satisfy the extra increase from this spike.

My proposal is to sort the inputs according to:
```golang
    1000 * vi - (ri * wi) > 1000 * vj - (rj * wj)
```
which is interpretted as the amount of satoshis remaining after subtracting the cost of spending via the input's witness type at this specific fee rate, scaled by 1000 to keep precision.

Since the non-witness contribution is identical for all input types, this can also be seen as the actual value that would be remaining, plus some contant. Additionally, the 0xfd'th input also has it's value slightly reduced since this is when the contribution from the varint is introduced. But since all outputs after it contain lower yield, no other sorting of the remaining inputs would have a positive yield either. I think this also generalizes to multiple such point decreases at fixed indices.

The remainder of the logic here works the same, and properly detects when this threshold is met and our sweep value starts to show negative ROI and we stop accepting new inputs to the sweep transaction. This should become more accurate as well once we move to a fixed size signature scheme, and the sizes are actual instead of worst-case.

Let me know your thoughts!

Reply:
Sorting by v - r*w is what I tried to do in the previous iteration of the sweeper. I didn't implement it correctly, as you pointed out then. When I started to fix the bug, I realized that the added weight calculation wasn't as easy as I thought (var int, witness yes/no). I abandoned it completely and resorted to sorting on output value only.

You point out some new things here. I like the reasoning that the varint penalty doesn't influence the optimal ordering. Also leaving the non-witness data out of the equation is nice.

There still is the case where you'd calculate with worst case weights, but in the end lower weights apply. 

To be clear, all of this isn't about sweeping negative yield inputs. That is prevented by the running weight calculation. It is just about sweeping sooner in specific cases.

Your proposal is an improvement, but I am in doubt whether I should implement it. It adds complexity because it requires a separate added weight calculation. And looking at the bigger topic of optimal sweep tx composition, I think neither approaches are optimal. For example: because of the sorting we publish a first tx containing all the big inputs. The leftovers aren't sweepable anymore after that. While if we would have saved some big inputs for the second tx, everything could have been swept.

Do you think adding the worst case weight calc. is worth it?

Reply:
> There still is the case where you'd calculate with worst case weights, but in the end lower weights apply.

Yeah unfortunately we can't properly account for this unless we try to sign each transaction, which would require a quadratic number of signatures. Even then, we can't count on an input's witness being the same size across multiple plausible sweeps unless we grind out signatures that create smaller witnesses :P May be something to consider for inputs right along the cusp of spendable tho...

> because of the sorting we publish a first tx containing all the big inputs. The leftovers aren't sweepable anymore after that. While if we would have saved some big inputs for the second tx, everything could have been swept.

My concern here is that because we do a linear probe until finding an input that decreases the total yield, we may eventually end up with outputs that are permanently stuck at the tail of the input list above a certain fee rate threshold.

The most obvious case I think this can occur is if two inputs are roughly equal in value, but the one with more input has a disproportionately greater witness size. If we sort only be value, the more spendable output will remain stuck behind the less spendable one, unless the fee rate drops signifcantly

As an example, consider two inputs:
 - i1: v1 = 1000, w1 = 109 (p2wkh)
 - i2: v2 = 1001, w2 = 322 (received htlc-success)

Then computing the yields:
 - y1 = 1000*1000 - r * 109
 - y2 = 1000*1001 - r * 322

Let's assume the fixed cost per input is C = 50k, solving for the maximum fee rate that each can be spent will add to the total yield of the txn gives:
 - r1 = 8715 sat/kw
 - r2 = 2953 sat/kw

If we choose to sort the outputs purely by value, i2 will always precede i1 in the final sorting. Furthermore, it will never make it onto a sweep until the sampled fee rate drops below r2. My concern is that one problem output like this could cause a backup of otherwise spendable outputs to pile up behind it

> While if we would have saved some big inputs for the second tx, everything could have been swept.

With the current itreation, I don't think we will do this though? We will always terminate after finding an input that decreases the yield. The question of choosing to take a loss to sweep those inputs is different IMO, goal here is to ensure we sweep all inputs that are sweepable on their own

> Do you think adding the worst case weight calc. is worth it?

Still unclear, to me changing the sorting is the most straight forward but would to great hear some different ideas on how to approach this

Reply:
At the same time, this sorting is  purely in memory, so we can always roll out a different heuristic at a later point. Just wanted to make sure we examine it, and if we want to circle back that’s probably fine as well :)

Reply:
Changes were not so bad. I implemented your proposal.